### PR TITLE
feat: 🎸 gif supported by SDWebImage

### PR DESCRIPTION
**Pull request by ShadowTourist**
Support gif by `SDWebImage`

### CODE REVIEW

**Comment by ShadowTourist** [comment]
|before|after|
|-|-|
|Screen Shot 2021-08-09 at 4 59 59 PM|Screen Shot 2021-08-09 at 5 05 22 PM|

**Comment by ShadowTourist** [comment]
Important: `Note: However, many differences behavior between iOS 13/14's is hard to fixup. Due to maintain issue, in the future release, we will drop the iOS 13 supports and always match SwiftUI 2.0's behavior.`

File: Sources/SAPCAI/UI/Common/SwiftUI/AvatarView.swift [hunk 1]
```diff
@@ -1,24 +1,18 @@
...
-                 })
-            .frame(width: 32, height: 32, alignment: .center)
+        if let url = URL(string: imageUrl) {
+            ImageViewWrapper(url: url) { $0 }
+                .scaledToFill()
+                .placeholder(Image(systemName: "person.crop.circle"))
+                .frame(width: 32, height: 32, alignment: .center)
+                .clipped()
+        } else {
+            Image(systemName: "icloud.slash")
```
Comment (MarcoEidinger) [inline Sources/SAPCAI/UI/Common/SwiftUI/AvatarView.swift:14]: is this new?

File: Sources/SAPCAI/UI/Common/UIKit/ImageViewWrapper.swift [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        switch self.state {
+        case .completed(image: let image):
+            return image.size
+        case .initial, .error:
+            return .zero
+        }
+    }
+}
+
+struct ImageViewWrapper<Content>: View where Content: View {
```
Comment (MarcoEidinger) [inline Sources/SAPCAI/UI/Common/UIKit/ImageViewWrapper.swift:27]: Please add a Preview for `ImageViewWrapper`

File: Sources/SAPCAI/UI/Common/UIKit/ImageViewWrapper.swift [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    var imageManager: ImageManager
+    let wrapper = UIImageViewWrapper()
+
+    func makeUIView(context: Self.Context) -> UIImageViewWrapper {
+        self.wrapper
+    }
+
+    func updateUIView(_ uiView: UIImageViewWrapper, context: UIViewRepresentableContext<ImageViewRepresentable>) {
+        uiView.imageView.contentMode = self.imageManager.contentMode
+        /// retry strategy missed
```
Comment (MarcoEidinger) [inline Sources/SAPCAI/UI/Common/UIKit/ImageViewWrapper.swift:109]: what does this comment mean?

**Review by MarcoEidinger — CHANGES_REQUESTED** [review]
Overall looks really good! 

I noticed that cache has a certain impact on memory. Lets add a section to the README to inform the app developer that app developer decides if/when image cache gets deleted. For example, for `onDisappear` for `AssistantView` the following is possible

```Swift
SDImageCache.shared.clearMemory()
SDImageCache.shared.clearDisk()
```

Also the README file has to be updated in the following areas
- https://github.com/SAP/cloud-sdk-ios-cai#dependencies
- https://github.com/SAP/cloud-sdk-ios-cai#image-related

**Review by MarcoEidinger — CHANGES_REQUESTED** [review]
Thank yo so much @ShadowTourist , that looks really great!!

@justinguo I assume that is also fine from your side?

Here is my last-minute nit-picking feedback. Once incorporated I will merge the PR into a new branch `devBreaking`.

Just in case you or Justin wonder why a new branch: there are already changes in dev branch which were not yet published. I might want to roll those out as a patch release but the gif support through SDWebImage is a minor/major release worth due to the breaking changes of switching package dependencies.

**Comment by justinguo** [comment]
Overall looks good to me. I tested and it works fine on device.

**Comment by justinguo** [comment]
@ShadowTourist Did you add the retry operation in `updateUIView()` method? Like if display failure view, can user tap the image again to retry the downloading and decoding?

**Comment by ShadowTourist** [comment]
> @ShadowTourist Did you add the retry operation in `updateUIView()` method? Like if display failure view, can user tap the image again to retry the downloading and decoding?

No, currently image view has no capability of interaction, so retry can be added by `SDWebImage`, there is a option called `SDWebImageRetryFailed` for image load failed. If we need add this for `ImageViewWrapper`?

**Review by MarcoEidinger — APPROVED** [review]

**Pull request merged**